### ISSUE TITLE
CA1812 should be reported for abstract types, unless instantiated

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#1159]
#### Analyzer package

Example: Microsoft.CodeQuality.Analyzers

#### Analyzer

CA1812: Avoid uninstantiated internal classes

#### Repro steps

1. Declare an internal class with the `abstract` modifier
1. Do not derive any types from the base class

```csharp
internal abstract class UninstantiatedBaseType
{
}
```

#### Expected behavior

CA1812 should be reported.

#### Actual behavior

CA1812 is not reported.

:memo: If the abstract type has one or more derived types, CA1812 should not be reported on the base class. If/when the derived types are removed in accordance with CA1812, a new warning will indicate that the user can now remove CA1812.

**Comment by 333fred** [comment]
Similar to #1163, this behavior is an explicit match with FxCop behavior, and unlike #1163, this would introduce a new warning where previously none existed. @dotnet/roslyn-analysis, thoughts on what we should do here? While I agree with the change in behavior, do we want to risk breaking people by reporting new warnings

**Comment by mavasani** [comment]
We should add an editorconfig option to allow end users to configure this behavior.

**Comment by Evangelink** [comment]
What about interfaces? If an `internal` interface is never implemented it shall have a similar behavior as un-inherited `abstract` classes, don't you think?

**Comment by Evangelink** [comment]
I also don't understand why we don't report on `struct` nor `enum`, is there any reason not to?

**Comment by JohanLarsson** [comment]
Bonus points for warning on code only used by tests but that is probably a separate issue.

**Comment by Evangelink** [comment]
@JohanLarsson If we go this road it would be nice to be able to categorize an assembly as main code/production code VS test code. I don't want to have some diagnostic reported on an internal test class used only by test class. This class could be either an helper directly in the test assembly or in some code of test utility/test infrastructure project.

**Comment by Evangelink** [comment]
@mavasani We have raised a couple of interesting cases here. In your opinion, which one(s) shall be handled?
